src/OrleansRuntime/Streams/PubSub/PubSubRendezvousGrain.cs

```diff
@@ -103,33 +101,57 @@ private static bool IsDeadProducer(IStreamProducerExtension producer)
 
         public async Task<ISet<PubSubSubscriptionState>> RegisterProducer(StreamId streamId, IStreamProducerExtension streamProducer)
         {
-            if (!IsActiveProducer(streamProducer))
-                throw new ArgumentException(String.Format("Trying to register non active IStreamProducerExtension: {0}", streamProducer.ToString()), "streamProducer");
-            
-            RemoveDeadProducers();
-            
-            var publisherState = new PubSubPublisherState(streamId, streamProducer);
-            State.Producers.Add(publisherState);
             counterProducersAdded.Increment();
-            counterProducersTotal.Increment();
-            LogPubSubCounts("RegisterProducer {0}", streamProducer);
-            await WriteStateAsync();
+            if (!IsActiveProducer(streamProducer))
+                throw new ArgumentException($"Trying to register non active IStreamProducerExtension: {streamProducer}", "streamProducer");
+
+            try
+            {
+                RemoveDeadProducers();
+
+                var publisherState = new PubSubPublisherState(streamId, streamProducer);
+                State.Producers.Add(publisherState);
+                LogPubSubCounts("RegisterProducer {0}", streamProducer);
+                await WriteStateAsync();
+                counterProducersTotal.Increment();
+            }
+            catch (Exception exc)
+            {
+                logger.Error(ErrorCode.Stream_RegisterProducerFailed, $"Failed to register a stream producer.  Stream: {streamId}, Producer: {streamProducer}, Exception: {exc}");
+                // Inconsistent state, deactivate grain.
```

Comment:
I'll modify comment as the term "inconsistent state" has a specific meaning in Orleans storage.